ohos: Improve the skeleton code of ovCompose access to OHOS unified rendering (Render Service)

sebuntin opened the pull request:
ohos: Add skeleton code for OHOS text measurement and rendering to prevent text rendering process from crashing

sq1007 committed:
ohos: Add skeleton code for OHOS text measurement and rendering

bulloo committed:
feat:add picture recorder for render node

Signed-off-by: wuxujie <yue5063657@163.com>

ywenjing commented on compose/foundation/foundation/src/commonMain/kotlin/androidx/compose/foundation/text/modifiers/TextStringSimpleNode.kt:
```diff
@@ -487,24 +488,24 @@ internal class TextStringSimpleNode(
             var currentParagraphHashCode = 0
             // region Tencent Code
-            if (drawInSkia || EnableIOSParagraph) {
+            if (drawInSkia || EnableIOSParagraph || platformTextDelegate == null) {
```
这里建议优化为配置

sebuntin replied:
增加 EnableOHOSParagraph 配置

sebuntin committed:
Merge pull request #1 from bulloo/wxjer_renderNode

feat:add picture recorder for render node

tencent-adm commented:
[![CLA assistant check](https://cla.opensource.tencent.com/pull/badge/signed)](https://cla.opensource.tencent.com/Tencent-TDS/ovCompose-multiplatform-core?pullRequest=14) <br/>All committers have signed the CLA.

sq1007 committed:
ohos: Implement OHComposeNativePaint and enhance AdaptiveCanvas with native paint synchronization

sq1007 committed:
ohos: Implement save, restore, and translate methods in OHNativeCanvasProxy

sq1007 committed:
ohos: Introduce NativeBasicShader and related gradient shaders for enhanced rendering

ywenjing committed:
Merge pull request #3 from ywenjing/ohos-compose-ywjing

Ohos compose ywjing

Commit:
refactor: Update enum references and improve rendering node structure

bulloo committed:
feat:wrapper OH_HiTrace

Signed-off-by: wuxujie <yue5063657@163.com>

cx478815108 commented on compose/ui/ui-arkui/src/ohosArm64Main/cpp/compose/src/main/cpp/compose/canvas/oh_native_canvas_layer_drawer.cpp:
```diff
@@ -0,0 +1,126 @@
+
+void OHRenderNodeDrawClipRect(float left, float top, float right, float bottom, const RenderNodeSaveState *saveState,
+                              BaseRenderNode *renderNodeForDrawing) {
+    ArkUI_RectShapeOption *shape = OH_ArkUI_RenderNodeUtils_CreateRectShapeOption();
```
clip 频率非常高，这个 shape 有办法全局复用吗？或者能否先判断当前的 renderNode 上是否已经存在了 shape ？

sebuntin replied:
可以考虑将这个shape提取到renderNode中，跟随renderNode的生命周期，只要renderNode不被销毁，shape也可以被这个renderNode一直复用

cx478815108 commented on compose/ui/ui-arkui/src/ohosArm64Main/cpp/compose/src/main/cpp/compose/render_node/oh_base_render_node.h:
```diff
@@ -0,0 +1,197 @@
+            return this;
+        }
+
+        BaseRenderNode* setTransform(float* matrix) {
```
下面已经有了 BaseRenderNode* setTransform(const float* matrix)，应该可以删除这个了` BaseRenderNode* setTransform(float* matrix)`

cx478815108 commented on compose/ui/ui-arkui/src/ohosArm64Main/cpp/compose/src/main/cpp/compose/render_node/oh_line_gradient_render_node.cpp:
```diff
@@ -0,0 +1,134 @@
+
+                // 释放绘制资源
+                OH_Drawing_CanvasDetachPen(canvas);
+                OH_Drawing_PathClose(path);
```
看上去少了 OH_Drawing_PathDetory(path) 的调用

cx478815108 commented on compose/ui/ui-arkui/src/ohosArm64Main/cpp/compose/src/main/cpp/compose/render_node/oh_line_render_node.cpp:
```diff
@@ -0,0 +1,145 @@
+
+                // 释放绘制资源
+                OH_Drawing_CanvasDetachPen(canvas);
+                OH_Drawing_PathClose(path);
```
`OH_Drawing_PathClose` 这个会释放 path 吗？看上去和 `OH_Drawing_PathDestory` 语义不一样

sebuntin replied:
这里应该是写错了，应该调用OH_Drawing_PathDestroy方法

cx478815108 commented on compose/ui/ui-arkui/src/ohosArm64Main/cpp/compose/src/main/cpp/compose/render_node/oh_rect_gradient_render_node.cpp:
```diff
@@ -0,0 +1,150 @@
+                auto *data = static_cast<RectGradientRenderNode *>(userData);
+                auto *canvas1 = OH_ArkUI_DrawContext_GetCanvas(context);
+                OH_Drawing_Canvas *canvas = reinterpret_cast<OH_Drawing_Canvas *>(canvas1);
+                OH_Drawing_ShaderEffect *shaderEffect = CreateShaderEffect(data->shader);
```
这里一定要创建 shaderEffect 吗？

cx478815108 commented on compose/ui/ui/src/ohosArm64Main/kotlin/androidx/compose/ui/platform/nativefoundation/AdaptiveCanvas.ohos.kt:
```diff
@@ -55,25 +59,34 @@ private inline fun CornerRadius.greaterThen(rhs: CornerRadius): Boolean {
+        val rawNativePaintHandle: OHComposeNativePaint_Handle? =
+            androidx_compose_ui_arkui_utils_OHNativeCanvasProxy_Paint(rawCanvasProxyHandle)
+        nativeCanvasProxy = OHNativeCanvasProxy(rawCanvasProxyHandle)
+        nativePaint = OHComposeNativePaint(rawNativePaintHandle)
```
github review 的时候，无法展开文件全文，这里不知道有没有在 dispsoe 里面去 释放nativeCanvasProxy 和  nativePaint

sebuntin replied:
需要在OHNativeCanvasProxy和OHNativePaint中增加cleaner，调用对应C++类的销毁方法

cx478815108 commented on compose/ui/ui/src/ohosArm64Main/kotlin/androidx/compose/ui/platform/nativefoundation/AdaptiveCanvas.ohos.kt:
```diff
@@ -55,25 +59,34 @@ private inline fun CornerRadius.greaterThen(rhs: CornerRadius): Boolean {
 
     override fun onPreDraw() {
         nativeCanvasProxy.beginDraw()
         LogPrintUtil.verbose("AdaptiveCanvas::beginDraw")
```
这里情况有点特殊，在 AdaptiveCanvas.ohos.kt 中需要先判断一个 debug 开关，开了才去 LogPrintUtil.verbose("AdaptiveCanvas::beginDraw")。否则这里字符串的创建是 100% 的，也会对整体的性能产生比较明显的影响

cx478815108 commented on compose/ui/ui/src/ohosArm64Main/kotlin/androidx/compose/ui/platform/nativefoundation/OHComposeNativePaint.kt:
```diff
@@ -0,0 +1,162 @@
+import androidx.compose.ui.graphics.StrokeJoin
+import platform.arkui.OH_Drawing_BlendMode
+
+internal fun BlendMode.asNativeBlendMode(): OH_Drawing_BlendMode {
```
建议加上 inline

cx478815108 commented on compose/ui/ui/src/ohosArm64Main/kotlin/androidx/compose/ui/platform/nativefoundation/OHComposeNativePaint.kt:
```diff
@@ -0,0 +1,162 @@
+class OHComposeNativePaint(val handle: OHComposeNativePaint_Handle?) {
+    fun sync(paint: Paint) {
+        // 实现与 OHComposeNativePaint 的同步逻辑
+        LogPrintUtil.verbose("OHComposeNativePaint::sync, compose paint: (${paint.toReadableString()})")
```
这里要小心，日志创建的字符串，是 100% 创建的

Commit:
ohos: Fix OverScroll animation issues by correctly parsing VSync signal timestamps..

Commit:
ohos: Provides native support for measuring, layout, and drawing of ovCompose text field on the ohos platform

sebuntin committed:
Merge pull request #7 from ywenjing/oh-compoe-ywenjing

ohos: Fix OverScroll animation issues by correctly parsing VSync sign…

Commit:
ohos: Add support for Debug configuration in build.gradle.kts and adjust compiler arguments accordingly

Commit:
ohos: Enhance AdaptiveCanvas with drawParagraph method and improve native canvas proxy resource management

sebuntin committed:
Merge pull request #8 from ywenjing/oh-compose-ywenjing

Oh compose ywenjing

Commit:
ohos: Refactor Paragraph handling to integrate spanStyle and placeholder management

Commit:
ohos: Refactor Paragraph handling to integrate spanStyle and placeholder management

Commit:
ohos: Refactor Paragraph handling to integrate spanStyle and placeholder management

bulloo committed:
Merge branch 'ohos-compose-1.6.1-20250919' into wxjer_trace

bulloo committed:
Merge pull request #5 from bulloo/wxjer_trace

feat:wrapper OH_HiTrace

Commit:
refactor: Update Paragraph class for improved style handling and memory management

Commit:
feat: Add isTraceEnabled property to TraceUtil for trace control

Commit:
refactor: Improve invalidate logic and enhance hash functions for better performance

Commit:
refactor: Enhance parameter passing and const correctness in canvas drawing functions

bulloo committed:
feat:implement drawingItem diff

Signed-off-by: wuxujie <yue5063657@163.com>

Commit:
fix: 解决LazyColumn组件的滑动问题

Commit:
feat: Add LayerSourceType metadata to distinguish LazyList item layers

- Add LayerSourceType enum (LAZY_LIST_ITEM, REGULAR)
- Extend OwnedLayer interface with getSourceType() method
- Update Owner.createLayer() API with sourceType parameter (default=REGULAR)
- Update OwnedLayerFactory.createLayer() with sourceType parameter
- Implement sourceType in ArkUIRenderNodeLayer with isLazyListItem() helper
- Update Placeable.placeWithLayer() and placeRelativeWithLayer() APIs
- Update NodeCoordinator and RootNodeOwner to propagate sourceType
- LazyListMeasuredItem now passes LAZY_LIST_ITEM when placing items

This enables OHOS platform to identify LazyList item layers.
fix: 解决LazyColumn组件的滑动问题

sebuntin committed:
Merge pull request #9 from bulloo/wxjer_diff

feat:implement drawingItem diff

Commit:
refactor: Update logging syntax to use lambda expressions for better readability

Commit:
refactor: Add some trace point and improve rendering performance

Commit:
feat: Add support for drawing oval, arc, image, and path shapes in canvas

feat: Add support for drawing oval, arc, and path shapes in canvas

Commit:
refactor: Improve rendering backend handling

Commit:
feat: Add support for drawing points and rendering text by converting text to SkBitmap

Commit:
feat: Implement clipping functionality for rectangles, round rectangles, and paths in the canvas

Commit:
feat: Enhance AdaptiveCanvas with clearClip, saveLayer, enableZ, and disableZ methods

Commit:
feat: Enhance AdaptiveCanvas with clearClip, saveLayer, enableZ, and disableZ methods

Commit:
feat: Enhance AdaptiveCanvas with clearClip, saveLayer, enableZ, and disableZ methods

Commit:
feat: Refactor drawing logic to incorporate translation adjustments and improve border style handling

Commit:
feat: Introduce AsyncTaskRenderNode for asynchronous rendering tasks and enhance color filter support

Commit:
feat: Add CircleGradientRenderNode and PathGradientRenderNode for enhanced gradient rendering capabilities

Commit:
feat: Implement OHDrawingPixelMapCache for optimized pixel map handling and reduce memory overhead
